File: sample/src/main/java/me/drakeet/multitype/sample/bilibili/BilibiliActivity.java
/*
 * Copyright 2016 drakeet. https://github.com/drakeet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.drakeet.multitype.sample.bilibili;

import android.os.Bundle;
import android.support.annotation.VisibleForTesting;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.GridLayoutManager.SpanSizeLookup;
import android.support.v7.widget.RecyclerView;
import android.view.Menu;
import android.view.MenuInflater;
import java.util.ArrayList;
import java.util.List;
import me.drakeet.multitype.MultiTypeAdapter;
import me.drakeet.multitype.sample.MenuBaseActivity;
import me.drakeet.multitype.sample.R;
import me.drakeet.multitype.sample.common.Category;
import me.drakeet.multitype.sample.common.CategoryItemViewBinder;

/**
 * @author drakeet
 */
public class BilibiliActivity extends MenuBaseActivity {

  private static final int SPAN_COUNT = 2;
  @VisibleForTesting List<Object> items;
  @VisibleForTesting MultiTypeAdapter adapter;


  private static class JsonData {

    private static final String PREFIX = "这是一条长长的达到两行的标题文字";

    private Post post00 = new Post(R.drawable.img_00, PREFIX + "post00");
    private Post post01 = new Post(R.drawable.img_01, PREFIX + "post01");
    private Post post10 = new Post(R.drawable.img_10, PREFIX + "post10");
    private Post post11 = new Post(R.drawable.img_11, PREFIX + "post11");

    Category category0 = new Category("title0");
    Post[] postArray = { post00, post01, post10, post11 };

    List<Post> postList = new ArrayList<>();


    {
      postList.add(post00);
      postList.add(post00);
      postList.add(post00);
      postList.add(post00);
      postList.add(post00);
      postList.add(post00);
    }
  }


  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_list);

    adapter = new MultiTypeAdapter();
    adapter.register(Category.class, new CategoryItemViewBinder());

    adapter.register(Post.class, new PostViewBinder());
    adapter.register(PostList.class, new HorizontalPostsViewBinder());

    RecyclerView recyclerView = findViewById(R.id.list);

    final GridLayoutManager layoutManager = new GridLayoutManager(this, SPAN_COUNT);
    SpanSizeLookup spanSizeLookup = new SpanSizeLookup() {
      @Override
      public int getSpanSize(int position) {
        Object item = items.get(position);
        return (item instanceof PostList || item instanceof Category) ? SPAN_COUNT : 1;
      }
    };
    layoutManager.setSpanSizeLookup(spanSizeLookup);
    recyclerView.setLayoutManager(layoutManager);
    int space = getResources().getDimensionPixelSize(R.dimen.normal_space);
    recyclerView.addItemDecoration(new PostItemDecoration(space, spanSizeLookup));

    recyclerView.setAdapter(adapter);

    JsonData data = new JsonData();
    items = new ArrayList<>();
    for (int i = 0; i < 10; i++) {
      /* You also could use Category as your CategoryItemContent directly */
      items.add(data.category0);
      items.add(data.postArray[0]);
      items.add(data.postArray[1]);
      items.add(data.postArray[2]);
      items.add(data.postArray[3]);
      items.add(data.postArray[0]);
      items.add(data.postArray[1]);
      items.add(new PostList(data.postList));
    }
    adapter.setItems(items);
    adapter.notifyDataSetChanged();
  }


  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    MenuInflater inflater = getMenuInflater();
    inflater.inflate(R.menu.menu_main, menu);
    return true;
  }
}

File: sample/src/main/kotlin/com/drakeet/multitype/sample/bilibili/BilibiliActivity.kt
/*
 * Copyright (c) 2016-present. Drakeet Xu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.drakeet.multitype.sample.bilibili

import android.os.Bundle
import android.view.Menu
import androidx.annotation.VisibleForTesting
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.GridLayoutManager.SpanSizeLookup
import androidx.recyclerview.widget.RecyclerView
import com.drakeet.multitype.MultiTypeAdapter
import com.drakeet.multitype.sample.MenuBaseActivity
import com.drakeet.multitype.sample.R
import com.drakeet.multitype.sample.common.Category
import com.drakeet.multitype.sample.common.CategoryItemViewBinder
import java.util.*

/**
 * @author Drakeet Xu
 */
class BilibiliActivity : MenuBaseActivity() {

  @VisibleForTesting
  internal lateinit var items: MutableList<Any>
  @VisibleForTesting
  internal lateinit var adapter: MultiTypeAdapter

  private class JsonData {

    private val post00 = Post(R.drawable.img_00, PREFIX + "post00")
    private val post01 = Post(R.drawable.img_01, PREFIX + "post01")
    private val post10 = Post(R.drawable.img_10, PREFIX + "post10")
    private val post11 = Post(R.drawable.img_11, PREFIX + "post11")

    internal var category0 = Category("title0")
    internal var postArray = arrayOf(post00, post01, post10, post11)

    internal var postList: MutableList<Post> = ArrayList()

    init {
      postList.add(post00)
      postList.add(post00)
      postList.add(post00)
      postList.add(post00)
      postList.add(post00)
      postList.add(post00)
    }

    companion object {
      private const val PREFIX = "这是一条长长的达到两行的标题文字"
    }
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_list)

    adapter = MultiTypeAdapter()
    adapter.register(CategoryItemViewBinder())

    adapter.register(PostViewBinder())
    adapter.register(HorizontalPostsViewBinder())

    val recyclerView = findViewById<RecyclerView>(R.id.list)

    val layoutManager = GridLayoutManager(this, SPAN_COUNT)
    val spanSizeLookup = object : SpanSizeLookup() {
      override fun getSpanSize(position: Int): Int {
        val item = items[position]
        return if (item is PostList || item is Category) SPAN_COUNT else 1
      }
    }
    layoutManager.spanSizeLookup = spanSizeLookup
    recyclerView.layoutManager = layoutManager
    val space = resources.getDimensionPixelSize(R.dimen.normal_space)
    recyclerView.addItemDecoration(PostItemDecoration(space, spanSizeLookup))

    recyclerView.adapter = adapter

    val data = JsonData()
    items = ArrayList()
    for (i in 0..9) {
      /* You also could use Category as your CategoryItemContent directly */
      items.add(data.category0)
      items.add(data.postArray[0])
      items.add(data.postArray[1])
      items.add(data.postArray[2])
      items.add(data.postArray[3])
      items.add(data.postArray[0])
      items.add(data.postArray[1])
      items.add(PostList(data.postList))
    }
    adapter.items = items
    adapter.notifyDataSetChanged()
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    val inflater = menuInflater
    inflater.inflate(R.menu.menu_main, menu)
    return true
  }

  companion object {
    private const val SPAN_COUNT = 2
  }
}
